lucene/core/src/java/org/apache/lucene/index/IndexWriter.java

```diff
@@ -3147,6 +3149,42 @@ public final boolean flushNextBuffer() throws IOException {
     }
   }
 
+  private MergePolicy.OneMerge updateSegmentInfosOnMergeFinish(MergePolicy.OneMerge merge, final SegmentInfos toCommit,
+                                                                AtomicReference<CountDownLatch> mergeLatchRef) {
+    return new MergePolicy.OneMerge(merge.segments) {
+      public void mergeFinished() throws IOException {
+        super.mergeFinished();
+        CountDownLatch mergeAwaitLatch = mergeLatchRef.get();
+        if (mergeAwaitLatch == null) {
+          // Commit thread timed out waiting for this merge and moved on. No need to manipulate toCommit.
```

Comment:
We need a stronger synchronization to make sure that we won't modify `toCommit` if `commitInternal` has stopped waiting for these merges.